### PR TITLE
Added preTransmission() and postTransmission(). Fixes #33

### LINKED ISSUE #33: How is RS485 Tx Enable controlled
I am fairly new to Arduinos and am looking for an easy to use Modbus RTU master.  With your ModbusMaster how is the Tx enable for the RS485 driver controlled?  I can't see anywhere in your library where this can be set up. I expected the ModbusMaster.begin to allow the TxEn signal to be allocated to any Arduino digital out pin. All I can see for .begin are parameters for choosing the Arduino serial port and the Modbus Id for the slave.

### PR DESCRIPTION
**Pull request by kintel**
This is a suggested API change to address dealing with RS-485 transceivers in half duplex mode (Issue #33).

I'm not married to this particular way of solving it, but it's tested to work on my end.

### CODE REVIEW

**Comment by 4-20ma** [comment]
Interested in the PR. Please fix conflicts and resubmit. Thanks.

**Comment by kintel** [comment]
Merged with master and resolved conflicts.

File: ModbusMaster.cpp [hunk 1]
```diff
@@ -183,6 +185,37 @@ void ModbusMaster::idle(void (*idle)())
   _idle = idle;
 }
 
+/**
+Set pre-transmission callback function.
+
+This function gets called just before a modbus message is sent over serial.
+Typical usage of this callback is to enable an RS485 transcieiver's
```
Comment (4-20ma) [inline ModbusMaster.cpp]: Spelling: "transcieiver's"

File: ModbusMaster.h [hunk 2]
```diff
@@ -259,6 +261,10 @@ class ModbusMaster
     
     // idle callback function; gets called during idle time between TX and RX
     void (*_idle)();
+    // preTransmission callback function; gets called before writing a modbus message
+    void (*_preTransmission)();
+    // postTransmission callback function; gets after a modbus message has been sent
```
Comment (4-20ma) [inline ModbusMaster.h]: "gets after" --> "gets called after"

File: ModbusMaster.h [hunk 3]
```diff
@@ -259,6 +261,10 @@ class ModbusMaster
     
     // idle callback function; gets called during idle time between TX and RX
     void (*_idle)();
+    // preTransmission callback function; gets called before writing a modbus message
```
Comment (4-20ma) [inline ModbusMaster.h]:
``` diff
- ...gets called before writing a modbus message
+ ...gets called before writing a Modbus message
```

File: ModbusMaster.cpp [hunk 4]
```diff
@@ -183,6 +185,37 @@ void ModbusMaster::idle(void (*idle)())
   _idle = idle;
 }
 
+/**
+Set pre-transmission callback function.
+
+This function gets called just before a modbus message is sent over serial.
```
Comment (4-20ma) [inline ModbusMaster.cpp]:
``` diff
- ...before a modbus message
+ ...before a Modbus message
```

File: ModbusMaster.cpp [hunk 5]
```diff
@@ -183,6 +185,37 @@ void ModbusMaster::idle(void (*idle)())
...
+*/
+void ModbusMaster::preTransmission(void (*preTransmission)())
+{
+  _preTransmission = preTransmission;
+}
+
+/**
+Set post-transmission callback function.
+
+This function gets called after a modbus message has finished sending
```
Comment (4-20ma) [inline ModbusMaster.cpp]:
``` diff
- ...after a modbus message
+ ...after a Modbus message
```

File: ModbusMaster.cpp [hunk 6]
```diff
@@ -183,6 +185,37 @@ void ModbusMaster::idle(void (*idle)())
...
+(i.e. after all data has been physically transmitted onto the serial
+bus).
+
+Typical usage of this callback is to enable an RS485 transcieiver's
```
Comment (4-20ma) [inline ModbusMaster.cpp]:
``` diff
- ...enable an RS485 transcieiver's
+ ...enable an RS485 transceiver's
```

File: ModbusMaster.cpp [hunk 7]
```diff
@@ -183,6 +185,37 @@ void ModbusMaster::idle(void (*idle)())
...
+Received Enable pin, and disable its Driver Enable pin.
```
Comment (4-20ma) [inline ModbusMaster.cpp]:
``` diff
- Received Enable pin...
+ Receiver Enable pin...
```

File: ModbusMaster.cpp [hunk 8]
```diff
@@ -672,6 +705,7 @@ uint8_t ModbusMaster::ModbusMasterTransaction(uint8_t u8MBFunction)
   while (_serial->read() != -1);
 
   // transmit request
+  if (_preTransmission) _preTransmission();
```
Comment (4-20ma) [inline ModbusMaster.cpp]: Let's maintain consistency with the _idle() function:

``` c++
if (_preTransmission)
{
  _preTransmission();
}
```

File: ModbusMaster.cpp [hunk 9]
```diff
@@ -683,6 +717,7 @@ uint8_t ModbusMaster::ModbusMasterTransaction(uint8_t u8MBFunction)
   
   u8ModbusADUSize = 0;
   _serial->flush();    // flush transmit buffer
+  if (_postTransmission) _postTransmission();
```
Comment (4-20ma) [inline ModbusMaster.cpp]: Let's maintain consistency with the _idle() function:

``` c++
if (_postTransmission)
{
  _postTransmission();
}
```

File: ModbusMaster.h [hunk 10]
```diff
@@ -45,7 +45,7 @@ Set to 1 to enable debugging features within class:
   - pin 5 cycles for each millisecond timeout during the Modbus response
 */
 #define __MODBUSMASTER_DEBUG__ (1)
-
+
```
Comment (4-20ma) [inline ModbusMaster.h]: Remove extraneous space.

**Comment by 4-20ma** [comment]
Thanks for the PR, though I think after I incorporated #43, it introduced new conflicts. I've marked up the PR with minor edits (spelling, capitalization, spacing, & convention). If you'll incorporate these and fix the merge conflict, I'm ready to merge.

**Comment by 4-20ma** [comment]
@kintel it occurred to me that we should include a simple example showing off the half-duplex feature. Do you have a simple bare-bones example to include?

**Comment by kintel** [comment]
@4-20ma I agree that an example would be nice. I can clean up one of my test sketches and submit that as a separate PR.

Thanks for being thorough btw. - it really helps knowing that someone actually read the code :)

**Comment by 4-20ma** [comment]
LGTM

**Comment by 4-20ma** [comment]
I look forward to a simplified example that shows off how to use the half-duplex feature. Thanks for implementing. Travis CI is now implemented (compiles each example included in `examples/`) so it will catch the obvious build errors automatically.

**Pull request merged**